src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,128 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private static final double INCREASING_COEFFICIENT = 10;
+    private static final int IMPORTANT_VALUE = 1;
+
+    private T[] elementData = (T[]) new Object[DEFAULT_SIZE];
+    private int size = 0;
+
+    private T[] grow(int oldSize) {
+        return (T[]) new Object[(int) (oldSize * INCREASING_COEFFICIENT)];
+    }
+
+    private T[] copiedList(T[] oldList, T[] newList) {
+        for (int i = 0; i < oldList.length; i++) {
+            newList[i] = oldList[i];
+        }
+        return newList;
+    }
+
+    private void addListInList(List<T> sourceList, T[] destList,
+                              int sizeOfSourceList, int sizeOfDestList) {
+        for (int i = sizeOfDestList, j = 0; j < sizeOfSourceList; j++, i++) {
+            destList[i] = sourceList.get(j);
+        }
+    }
+
+    private void droppedValue(T[] elementData, int index) {
+        int numMoved = size - index - IMPORTANT_VALUE;
+        System.arraycopy(elementData, index + IMPORTANT_VALUE, elementData, index, numMoved);
+        elementData[size - IMPORTANT_VALUE] = null;
+        size -= IMPORTANT_VALUE;
+    }
 
     @Override
     public void add(T value) {
-
+        if (size < elementData.length) {
+            elementData[size] = value;
+        } else {
+            elementData = copiedList(elementData, grow(elementData.length));
+            elementData[size] = value;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= size + IMPORTANT_VALUE || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index passed to the method is invalid");
+        }
+        if (DEFAULT_SIZE > elementData.length || size < elementData.length) {
+            System.arraycopy(elementData, index, elementData,
+                    index + IMPORTANT_VALUE, size - index);
+            elementData[index] = value;
+        } else {
+            T[] widerList = copiedList(elementData, grow(elementData.length));
+            System.arraycopy(elementData, index, widerList, index + IMPORTANT_VALUE, size - index);
```

Comment:
can we move this part to method grow entirely?

Reply:
I don't think so. Because I have already realized other type of System.arraycopya in grow method